bayes_kit/ensemble.py

```diff
@@ -1,68 +1,91 @@
+        a: bounds on proposal (default 2)
+        walkers: an even number of walkers to use (default dimensionality of `model * 2`)
+        init: `walker` x `dimensio`n array of initial positions (defaults to standard normal)
+
+        Throws:
+        ValueError: if `a` is provided and not >= 1, `walker`s is provided and not strictly positive and even,
+        or if the `init` is provided and is not an `NDArray` of shape `walker` x `dimension`
+        """
+        self._model = model
+        self._dim = self._model.dims()
+        if a != None and np.float64(a) < 1:
+            raise ValueError(f"stretch bound must be greater than or equal to 1; found {a=}")
+        self._a = np.float64(a or 2.0)
+        self._sqrt_a = np.sqrt(np.float64(a))
+        self._inv_sqrt_a = 1 / self._sqrt_a
+        self._walkers = np.int64(walkers or 2 * self._dim)
+        if self._walkers < 2 or self._walkers % 2 != 0:
+            raise ValueError(f"walkers must be strictly positive, even integer; found {walkers=}")
+        self._halfwalkers = self._walkers // 2
+        self._drawshape = (int(self._walkers), self._dim)
+        self._thetas = np.asarray(init or np.random.normal(size=self._drawshape))
+        if self._thetas.shape != self._drawshape:
+            raise ValueError(f"init must be shape of draw {self._drawshape}; found {self._thetas.shape=}")
+        self._firsthalf = range(0, int(self._halfwalkers))
+        self._secondhalf = range(int(self._halfwalkers), int(self._walkers))
 
-    # def __iter__(self):
-    #     return self
+    def __iter__(self) -> Iterator[Sample]:
+        return self
 
-    # def __next__(self):
-    #     return self.sample
+    def __next__(self) -> Sample:
+        return self.sample()
 
-    # def draw_z(self):
-    #     """Return random draw z in (1/a, a) with p(z) propto 1 / sqrt(z)"""
-    #     return np.square(np.random.uniform(self._inv_sqrt_a, self._sqrt_a))
+    def draw_z(self) -> Sample:
+        """Return random draw z in (1/a, a) with p(z) propto 1 / sqrt(z)"""
+        return np.asarray(np.square(np.random.uniform(self._inv_sqrt_a, self._sqrt_a)))
 
-    # def stretch_move(self, theta_k: NDarray[np.float64], theta_j: NDarray[np.float64]):
-    #     z = self.draw_z()
-    #     theta_star = theta_j + z * (theta_k - theta_j)  # (1 - z) * theta_j + z * theta_k
-    #     log_q = (self._dims - 1) * np.log(z) + self._model.log_density(theta_star) - self._model.log_density(theta_k)
-    #     if np.log(np.random.uniform()) < log_q:
-    #         return theta_star
-    #     return theta_k
+    def stretch_move(self, theta_k: NDArray[np.float64], theta_j: NDArray[np.float64]) -> Sample:
+        z = self.draw_z()
+        theta_star = np.asarray(theta_j + z * (theta_k - theta_j))  # (1 - z) * theta_j + z * theta_k
+        print(f"{theta_k=}  {theta_j=}  {z=}  {theta_star=}")
+        log_q = (self._dim - 1) * np.log(z) + self._model.log_density(theta_star) - self._model.log_density(theta_k)
```

Comment:
Would it be worth caching the log-density of each `theta_k`? Maybe in a `self._lp_thetas` or similar?

Reply:
Yes, absolutely.  That's most of the compute time.  I'll cache and get rid of superfluous prints (D'oh, again with feeling).